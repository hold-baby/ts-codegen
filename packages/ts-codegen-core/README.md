# TS Codegen Core

[![Build Status](https://img.shields.io/travis/reeli/ts-codegen.svg?style=flat-square&branch=master)](https://travis-ci.org/reeli/ts-codegen)
[![codecov](https://codecov.io/gh/reeli/ts-codegen/branch/master/graph/badge.svg?style=flat-square)](https://codecov.io/gh/reeli/ts-codegen)
[![License](https://img.shields.io/npm/l/@ts-tool/ts-codegen.svg?style=flat-square)](https://npmjs.org/package/@ts-tool/ts-codegen)

Generate ts code from swagger.

# 开始

1. `npm install @ts-codegen-core`
