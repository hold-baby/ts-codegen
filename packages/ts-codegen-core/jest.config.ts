export default {
  testEnvironment: "node",
  modulePaths: ["<rootDir>"],
  moduleNameMapper: {
    "^@ts-tool/([^/]+)$": "<rootDir>/src/index.ts",
  },
  coverageDirectory: "coverage",
  testRegex: ".*/__tests__/.+\\.(generator|test|spec)\\.(ts|tsx)$",
  collectCoverage: true,
  collectCoverageFrom: ["src/**/*.{ts,tsx}", "!src/__types__/**", "!src/bootstrap.ts", "!src/index.ts"],
  transform: {
    "^.+\\.tsx?$": "ts-jest",
  },
};
