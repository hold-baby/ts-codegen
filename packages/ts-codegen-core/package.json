{
  "name": "@ts-tool/ts-codegen-core",
  "version": "3.3.0",
  "main": "dist/lib/index.js",
  "module": "dist/module/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist",
    "README.md"
  ],
  "scripts": {
    "start": "ts-node -r tsconfig-paths/register src/bootstrap.ts"
  },
  "dependencies": {
    "axios": "0.21.x",
    "js-yaml": "4.0.x",
    "lodash": "4.17.x",
    "prettier": "*"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "reeli",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/reeli/ts-codegen/issues"
  },
  "homepage": "https://github.com/reeli/ts-codegen#readme"
}
