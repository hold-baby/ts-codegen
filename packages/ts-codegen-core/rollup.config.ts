import path from "path";
// @ts-ignore
import rollupBabel from "rollup-plugin-babel";
// @ts-ignore
import rollupTypeScript from "rollup-plugin-typescript";

const pkg = require(path.join(process.cwd(), "package.json"));

module.exports = {
  input: "src/index.ts",
  output: [
    {
      dir: "dist/lib",
      format: "cjs",
    },
    {
      dir: "dist/module",
      format: "es",
    },
  ],
  external: [
    // @ts-ignore
    ...Object.keys(process.binding("natives")),
    ...Object.keys(pkg.dependencies || {}),
    ...Object.keys(pkg.peerDependencies || {}),
  ],
  plugins: [
    rollupTypeScript({
      target: "es5",
      module: "es6",
    }),
    rollupBabel({
      plugins: ["babel-plugin-pure-calls-annotation"],
      exclude: "node_modules/**",
      extensions: [".js", ".jsx", ".ts", ".tsx"],
    }),
  ],
};
