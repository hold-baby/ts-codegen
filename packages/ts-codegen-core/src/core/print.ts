import { IClientConfig, RequestType, ScanOptions } from "../__types__/types";
import { IStore, DeclKinds } from "./createRegister";
import { prettifyCode, objToTypeStr } from "../utils/common";
import { sortBy, isEmpty, compact, keys, mapValues, isString } from "lodash";
import { CustomType } from "./Type";
import { DEFAULT_SERVICE_NAME_IN_HEADER } from "..";

export const printOutputs = (
  clientConfigs: IClientConfig[],
  decls: IStore["decls"],
  requestCreateMethod: string = "createRequest",
  options?: ScanOptions,
) => {
  return prettifyCode(`${printRequest(clientConfigs, requestCreateMethod, options)} \n\n ${printTypes(decls)}`);
};

const hasRequestBody = (TReq: IClientConfig["TReq"]) => TReq?.requestBody || (TReq || {})["requestBody?"];

const printRequest = (clientConfigs: IClientConfig[], requestCreateMethod: string, options?: ScanOptions): string => {
  const configs = sortBy(clientConfigs, (o) => o.operationId);

  return configs
    .map((v) => {
      const toUrl = () => `url: \`${v.url}\`,`;
      const toMethod = () => `method: "${v.method}",`;
      const toRequestBody = () => {
        if (!isEmpty(v.bodyParams)) {
          return `data: ${v.bodyParams!.length > 1 ? `{${v.bodyParams!.join(",")}}` : v.bodyParams![0]},`;
        }
        return v.contentType && hasRequestBody(v.TReq) ? "data: requestBody," : "";
      };
      const toQueryParams = () => {
        const params = toRequestParams(v.queryParams);
        return params ? `params: ${params},` : "";
      };
      const toHeaders = () => {
        if (v.contentType && options?.withServiceNameInHeader) {
          return `headers: { 'Content-Type': '${v.contentType}', '${
            isString(options?.withServiceNameInHeader)
              ? options?.withServiceNameInHeader
              : DEFAULT_SERVICE_NAME_IN_HEADER
          }': serviceName },`;
        }

        if (v.contentType) {
          return `headers: { 'Content-Type': '${v.contentType}' },`;
        }

        if (options?.withServiceNameInHeader) {
          return `headers: { '${
            isString(options?.withServiceNameInHeader)
              ? options?.withServiceNameInHeader
              : DEFAULT_SERVICE_NAME_IN_HEADER
          }': serviceName },`;
        }

        return "";
      };

      const toGenerators = () => {
        const TReq = generateTReq(v.TReq);
        const TResp = v.TResp?.toType(false);

        if (!TReq && !TResp) {
          return "";
        }
        if (!TResp) {
          return `<${TReq}>`;
        }
        return `<${TReq}, ${TResp}>`;
      };
      const toRequestInputs = () => {
        const getRequestBody = () => {
          if (!isEmpty(v.bodyParams)) {
            return v.bodyParams!;
          }
          return v.contentType && hasRequestBody(v.TReq) ? ["requestBody"] : "";
        };
        const list = compact([...v.pathParams, ...v.queryParams, ...getRequestBody()]);
        return isEmpty(list) ? "" : toRequestParams(list);
      };

      return `
${addComments(v, options?.withComments)}
export const ${v.operationId} = ${requestCreateMethod}${toGenerators()}("${
        v.operationId
      }", (${toRequestInputs()}) => ({${toUrl()}${toMethod()}${toRequestBody()}${toQueryParams()}${toHeaders()}})
);
`;
    })
    .join("");
};

const addComments = (v: IClientConfig, withComments?: boolean) => {
  const summaryComment = withComments && v.summary ? `* ${v.summary}` : "";
  const deprecatedComment = v.deprecated ? `* @deprecated ${v.operationId}` : "";
  const comments = [summaryComment, deprecatedComment].filter((c) => !!c);

  return isEmpty(comments)
    ? ""
    : `/**
  ${comments.join("\\n")}
  */`;
};

const printTypes = (decls: IStore["decls"]): string => {
  return keys(decls)
    .sort()
    .map((k) => {
      const expr = decls[k].kind === DeclKinds.type ? "=" : "";
      const semi = decls[k].kind === DeclKinds.type ? ";" : "";
      return `export ${decls[k].kind} ${decls[k].name} ${expr} ${decls[k].type.toType()}${semi}`;
    })
    .join("\n\n");
};

const generateTReq = (TReq: IClientConfig["TReq"]) => {
  if (isEmpty(TReq)) {
    return;
  }

  function gen(obj: IClientConfig["TReq"]): string {
    return objToTypeStr(
      mapValues(obj, (v) => {
        if (!v.toType) {
          return gen(v as RequestType);
        }
        return (v as CustomType).toType(false);
      }),
    );
  }

  return gen(TReq);
};

const toRequestParams = (data: string[]) => (!isEmpty(data) ? `{\n ${data.join(",\n")} \n}` : undefined);
