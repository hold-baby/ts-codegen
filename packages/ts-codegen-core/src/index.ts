export * from "./core";
export * from "./utils";
export * from "./constants";
export * from "./codegen";
export * from "./__types__/OpenAPI";
export * from "./__types__/types";
