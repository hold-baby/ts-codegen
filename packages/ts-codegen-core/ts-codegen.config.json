{
  "outputFolder": "clients",
  "requestCreateLib": "../examples/utils/createRequest",
  "requestCreateMethod": "createRequest",
  "apiSpecsPaths": [
    {
      "path": "https://petstore.swagger.io/v2/swagger.json",
      "name": "PetStoreService1"
    },
    {
      "path": "https://petstore.swagger.io/v2/swagger.yaml",
      "name": "PetStoreService2"
    },
    {
      "path": "./examples/swagger.json",
      "name": "SwaggerService"
    },
    {
      "path": "./examples/openapi.json",
      "name": "OpenApiService"
    },
    {
      "path": "./examples/openapi1.json",
      "name": "OpenApiService1"
    },
    {
      "path": "./examples/demo.yaml",
      "name": "DemoService"
    }
  ],
  "options": {
    "withComments": true,
    "typeWithPrefix": true,
    "backwardCompatible": true,
    "withServiceNameInHeader": true
  }
}
