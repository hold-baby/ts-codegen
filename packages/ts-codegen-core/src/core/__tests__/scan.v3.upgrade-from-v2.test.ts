import swaggerV3 from "../../../examples/openapi1.json";
import { scan } from "@ts-tool/ts-codegen-core";
import { printOutputs } from "../print";

describe("swagger v3", () => {
  it("should handle basic schemas correctly", () => {
    const { clientConfigs, decls } = scan(swaggerV3 as any);
    expect(printOutputs(clientConfigs, decls)).toMatchSnapshot();
  });
});
