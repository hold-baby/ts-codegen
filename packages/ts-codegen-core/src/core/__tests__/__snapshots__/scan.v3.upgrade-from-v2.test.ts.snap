// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swagger v3 should handle basic schemas correctly 1`] = `
"export const addPet = createRequest<{
  requestBody: Pet;
}>(\\"addPet\\", ({ requestBody }) => ({
  url: \`/api/v2/pet\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createItem = createRequest<
  {
    requestBody?: Item;
  },
  Message
>(\\"createItem\\", ({ requestBody }) => ({
  url: \`/api/v2/items\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUser = createRequest<{
  requestBody: User;
}>(\\"createUser\\", ({ requestBody }) => ({
  url: \`/api/v2/user\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithArrayInput = createRequest<{
  requestBody: User[];
}>(\\"createUsersWithArrayInput\\", ({ requestBody }) => ({
  url: \`/api/v2/user/createWithArray\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const createUsersWithListInput = createRequest<{
  requestBody: User[];
}>(\\"createUsersWithListInput\\", ({ requestBody }) => ({
  url: \`/api/v2/user/createWithList\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const deleteAttachmentUsingDelete = createRequest<{
  id: string;
}>(\\"deleteAttachmentUsingDelete\\", ({ id }) => ({ url: \`/api/v2/\${id}\`, method: \\"DELETE\\" }));

export const deleteOrder = createRequest<{
  orderId: number;
}>(\\"deleteOrder\\", ({ orderId }) => ({ url: \`/api/v2/store/order/\${orderId}\`, method: \\"DELETE\\" }));

export const deletePet = createRequest<{
  petId: number;
}>(\\"deletePet\\", ({ petId }) => ({ url: \`/api/v2/pet/\${petId}\`, method: \\"DELETE\\" }));

export const deleteUser = createRequest<{
  username: string;
}>(\\"deleteUser\\", ({ username }) => ({ url: \`/api/v2/user/\${username}\`, method: \\"DELETE\\" }));

export const downloadUsingGet = createRequest<
  {
    id: string;
  },
  Resource
>(\\"downloadUsingGet\\", ({ id }) => ({ url: \`/api/v2/\${id}\`, method: \\"GET\\" }));

export const findBookByIdUsingGet = createRequest<
  {
    id: string;
  },
  BookDetailVo
>(\\"findBookByIdUsingGet\\", ({ id }) => ({ url: \`/api/v2/book/\${id}\`, method: \\"GET\\" }));

export const findPetsByStatus = createRequest<
  {
    status: (keyof typeof FindPetsByStatusStatus)[];
  },
  Pet[]
>(\\"findPetsByStatus\\", ({ status }) => ({
  url: \`/api/v2/pet/findByStatus\`,
  method: \\"GET\\",
  params: {
    status,
  },
}));

/**
 * @deprecated findPetsByTags
 */
export const findPetsByTags = createRequest<
  {
    tags: string[];
  },
  Pet[]
>(\\"findPetsByTags\\", ({ tags }) => ({
  url: \`/api/v2/pet/findByTags\`,
  method: \\"GET\\",
  params: {
    tags,
  },
}));

export const getAgencyResponseListUsingGet = createRequest<
  {
    addressType: keyof typeof GetAgencyResponseListUsingGetAddressType;
  },
  Resource[]
>(\\"getAgencyResponseListUsingGet\\", ({ addressType }) => ({
  url: \`/api/v2/agency-responses\`,
  method: \\"GET\\",
  params: {
    addressType,
  },
}));

export const getAgencyResponseUsingGet = createRequest<
  {
    addressType: keyof typeof GetAgencyResponseUsingGetAddressType;
    responseDate: string;
  },
  Resource
>(\\"getAgencyResponseUsingGet\\", ({ responseDate, addressType }) => ({
  url: \`/api/v2/agency-response\`,
  method: \\"GET\\",
  params: {
    responseDate,
    addressType,
  },
}));

export const getDocumentByIdUsingGet = createRequest<
  {
    documentId: string;
    from?: CarTypeEnum;
  },
  DocumentVo
>(\\"getDocumentByIdUsingGet\\", ({ documentId, from }) => ({
  url: \`/api/v2/documents/\${documentId}/doc\`,
  method: \\"GET\\",
  params: {
    from,
  },
}));

export const getHealthCheckUsingHead = createRequest<
  {
    flight_no: string;
  },
  DocumentVo
>(\\"getHealthCheckUsingHead\\", ({ flight_no }) => ({ url: \`/api/v2/health-check/\${flight_no}\`, method: \\"HEAD\\" }));

export const getInventory = createRequest<undefined, { [key: string]: number }>(\\"getInventory\\", () => ({
  url: \`/api/v2/store/inventory\`,
  method: \\"GET\\",
}));

export const getItems = createRequest<
  {
    limit?: number;
    page?: number;
    sort?: string[];
  },
  Items
>(\\"getItems\\", ({ page, limit, sort }) => ({
  url: \`/api/v2/items\`,
  method: \\"GET\\",
  params: {
    page,
    limit,
    sort,
  },
}));

export const getOrderById = createRequest<
  {
    orderId: number;
  },
  Order
>(\\"getOrderById\\", ({ orderId }) => ({ url: \`/api/v2/store/order/\${orderId}\`, method: \\"GET\\" }));

export const getPetById = createRequest<
  {
    petId: number;
  },
  Pet
>(\\"getPetById\\", ({ petId }) => ({ url: \`/api/v2/pet/\${petId}\`, method: \\"GET\\" }));

export const getScheduleDetailsByDateUsingGet = createRequest<
  {
    roleId?: string;
    scheduleDate: number;
  },
  ScheduleVo[]
>(\\"getScheduleDetailsByDateUsingGet\\", ({ scheduleDate, roleId }) => ({
  url: \`/api/v2/schedules\`,
  method: \\"GET\\",
  params: {
    scheduleDate,
    roleId,
  },
}));

export const getTeamListByUsingGet = createRequest<{
  limit?: number;
  offset?: number;
}>(\\"getTeamListByUsingGet\\", ({ offset, limit }) => ({
  url: \`/api/v2/teams\`,
  method: \\"GET\\",
  params: {
    offset,
    limit,
  },
}));

export const getUserByName = createRequest<
  {
    username: string;
  },
  User
>(\\"getUserByName\\", ({ username }) => ({ url: \`/api/v2/user/\${username}\`, method: \\"GET\\" }));

export const getUserListByUsingGet = createRequest<{
  limit?: number;
  offset?: number;
}>(\\"getUserListByUsingGet\\", ({ offset, limit }) => ({
  url: \`/api/v2/users\`,
  method: \\"GET\\",
  params: {
    offset,
    limit,
  },
}));

export const loginUser = createRequest<
  {
    password: string;
    username: string;
  },
  string
>(\\"loginUser\\", ({ username, password }) => ({
  url: \`/api/v2/user/login\`,
  method: \\"GET\\",
  params: {
    username,
    password,
  },
}));

export const logoutUser = createRequest(\\"logoutUser\\", () => ({ url: \`/api/v2/user/logout\`, method: \\"GET\\" }));

export const placeOrder = createRequest<
  {
    requestBody: Order;
  },
  Order
>(\\"placeOrder\\", ({ requestBody }) => ({
  url: \`/api/v2/store/order\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updateBookByIdUsingPut = createRequest<{
  id: string;
  requestBody: UpdateBookRequest;
}>(\\"updateBookByIdUsingPut\\", ({ id, requestBody }) => ({
  url: \`/api/v2/book/\${id}\`,
  method: \\"PUT\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePet = createRequest<{
  requestBody: Pet;
}>(\\"updatePet\\", ({ requestBody }) => ({
  url: \`/api/v2/pet\`,
  method: \\"PUT\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const updatePetWithForm = createRequest<{
  petId: number;
  requestBody?: { name?: string; status?: string };
}>(\\"updatePetWithForm\\", ({ petId, requestBody }) => ({
  url: \`/api/v2/pet/\${petId}\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/x-www-form-urlencoded\\" },
}));

export const updateUser = createRequest<{
  requestBody: User;
  username: string;
}>(\\"updateUser\\", ({ username, requestBody }) => ({
  url: \`/api/v2/user/\${username}\`,
  method: \\"PUT\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"application/json\\" },
}));

export const uploadAttachmentUsingPost = createRequest<
  {
    requestBody?: { attachment: Blob };
  },
  AttachmentBo
>(\\"uploadAttachmentUsingPost\\", ({ requestBody }) => ({
  url: \`/api/v2\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export const uploadFile = createRequest<
  {
    petId: number;
    requestBody?: { additionalMetadata?: string; file?: Blob };
  },
  ApiResponse
>(\\"uploadFile\\", ({ petId, requestBody }) => ({
  url: \`/api/v2/pet/\${petId}/uploadImage\`,
  method: \\"POST\\",
  data: requestBody,
  headers: { \\"Content-Type\\": \\"multipart/form-data\\" },
}));

export interface ApiResponse {
  code?: number;
  message?: string;
  type?: string;
}

export interface AttachmentBo {
  authorName?: string;
  createdDate?: number;
  fileName?: string;
  id?: string;
  mimeType?: string;
  path?: string;
}

export interface BookDetailVo {
  attachment?: ScheduleVo;
  authorName?: string;
  createdDate?: number;
  fileName?: string;
  id?: string;
  mimeType?: string;
  path?: string;
}

export interface BookVo {
  address?: string;
  price?: string;
}

export interface BookingResponse {
  data: DocumentVo;
  errors?: ErrorInfo[];
}

export enum CarTypeEnum {
  \\"BMW\\" = \\"BMW\\",
  \\"Audi\\" = \\"Audi\\",
  \\"BYD\\" = \\"BYD\\",
}

export interface Category {
  id?: number;
  name?: string;
}

export interface Dictionary {
  category?: Category;
  dictionary1?: { age: number; name: string; [key: string]: string | number };
  dictionary2?: { age: string; name: string; [key: string]: string };
  dictionary3?: { age?: number; name?: string; [key: string]: any };
  dictionary4?: { [key: string]: any };
  dictionary?: { [key: string]: { code?: number; text?: string } };
  id?: number;
  name: string;
  photoUrls: string[];
  status?: keyof typeof DictionaryStatus;
  tags?: Tag[];
}

export enum DictionaryStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export interface DocumentVo {
  attachment?: BookDetailVo;
  authorName?: string;
  createdDate?: number;
  id?: string;
  note?: string;
  title?: string;
}

export interface Error {
  code?: number;
  field?: string;
  message?: string;
}

export interface ErrorInfo {
  errorMessage?: string;
}

export type Errors = Error[];

export interface File {
  absolute?: boolean;
  absoluteFile?: File;
  absolutePath?: string;
  canonicalFile?: File;
  canonicalPath?: string;
  directory?: boolean;
  executable?: boolean;
  file?: boolean;
  freeSpace?: number;
  hidden?: boolean;
  lastModified?: number;
  name?: string;
  parent?: string;
  parentFile?: File;
  path?: string;
  readable?: boolean;
  totalSpace?: number;
  usableSpace?: number;
  writable?: boolean;
}

export enum FindPetsByStatusStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export enum GetAgencyResponseListUsingGetAddressType {
  \\"AAA\\" = \\"AAA\\",
  \\"BBB\\" = \\"BBB\\",
}

export enum GetAgencyResponseUsingGetAddressType {
  \\"AAA\\" = \\"AAA\\",
  \\"BBB\\" = \\"BBB\\",
}

export interface InputStream {
  [key: string]: any;
}

export interface Item {
  color?: string;
  id?: string;
  name?: string;
  price?: string;
}

export type Items = Item[];

export interface Message {
  message?: string;
}

export interface Order {
  complete?: boolean;
  id?: number;
  petId?: number;
  quantity?: number;
  shipDate?: string;
  status?: keyof typeof OrderStatus;
}

export enum OrderStatus {
  \\"placed\\" = \\"placed\\",
  \\"approved\\" = \\"approved\\",
  \\"delivered\\" = \\"delivered\\",
}

export interface Pet {
  category?: Category;
  id?: number;
  name: string;
  photoUrls: string[];
  status?: keyof typeof PetStatus;
  tags?: Tag[];
}

export enum PetStatus {
  \\"available\\" = \\"available\\",
  \\"pending\\" = \\"pending\\",
  \\"sold\\" = \\"sold\\",
}

export interface Resource {
  description?: string;
  file?: File;
  filename?: string;
  inputStream?: InputStream;
  open?: boolean;
  readable?: boolean;
  uri?: Uri;
  url?: Url;
}

export interface ScheduleVo {
  contentType?: ScheduleVoContentType;
  reason?: ScheduleVoReason;
  schedules?: BookVo[][];
  shiftId?: string;
  team?: string;
}

export type ScheduleVoContentType = \\"01\\" | \\"02\\" | \\"03\\" | \\"04\\" | \\"05\\" | \\"06\\";

export type ScheduleVoReason = \\"1\\" | \\"3\\" | \\"4\\" | \\"5\\" | \\"6\\";

export interface Tag {
  id?: number;
  name?: string;
}

export interface UpdateBookRequest {
  birthCountry?: string;
  citizenship?: string;
  dateOfBirth?: number;
  employmentStatus?: string;
  ethnicity?: string;
  gender?: string;
  idNumber?: string;
  idType?: string;
  roleId?: string;
  spokenLanguage?: string[];
}

export interface Uri {
  absolute?: boolean;
  authority?: string;
  fragment?: string;
  host?: string;
  opaque?: boolean;
  path?: string;
  port?: number;
  query?: string;
  rawAuthority?: string;
  rawFragment?: string;
  rawPath?: string;
  rawQuery?: string;
  rawSchemeSpecificPart?: string;
  rawUserInfo?: string;
  scheme?: string;
  schemeSpecificPart?: string;
  userInfo?: string;
}

export interface Url {
  authority?: string;
  content?: { [key: string]: any };
  defaultPort?: number;
  deserializedFields?: UrlStreamHandler;
  file?: string;
  host?: string;
  path?: string;
  port?: number;
  protocol?: string;
  query?: string;
  ref?: string;
  serializedHashCode?: number;
  userInfo?: string;
}

export interface UrlStreamHandler {
  [key: string]: any;
}

export interface User {
  email?: string;
  firstName?: string;
  id?: number;
  lastName?: string;
  password?: string;
  phone?: string;
  userStatus?: number;
  username?: string;
}

export interface ValidationError {
  errors?: Errors;
  message?: string;
}
"
`;
